bitbots_ros_control/src/wolfgang_hardware_interface.cpp

```diff
@@ -8,102 +7,298 @@ namespace bitbots_ros_control {
+            driver->setTools(model_number_specified_16, id);
+            float mounting_offset;
+            dxl_nh.param<float>("mounting_offset", mounting_offset, 0.0);
+            float joint_offset;
+            dxl_nh.param<float>("joint_offset", joint_offset, 0.0);
+            servos_on_port.push_back(std::make_tuple(id, name, mounting_offset, joint_offset));
+          } else {
+            if (!only_pressure_ && !only_imu_) {
+              ROS_WARN("Could not identify device for ID %d", id);
+            }
+          }
+          pinged.push_back(name);
+        }
+      }
+    }
+    // create a servo bus interface if there were servos found on this bus
+    if (servos_on_port.size() > 0) {
+      ServoBusInterface *interface = new ServoBusInterface(driver, servos_on_port);
+      interfaces_on_port.push_back(interface);
+      servo_interface_.addBusInterface(interface);
+    }
+    // add vector of interfaces on this port to overall collection of interfaces
+    interfaces_.push_back(interfaces_on_port);
   }
-  if(success) {
-    speakError(speak_pub_, "ros control startup successful");
-  }else{
-    speakError(speak_pub_, "error starting ros control");
+
+  if (pinged.size() != dxl_devices.size()) {
+    // when we only have 1 or two devices its only the core
+    if (pinged.empty() || pinged.size() == 1  || pinged.size() == 2) {
+      ROS_ERROR("Could not start ros control. Power is off!");
+      speakError(speak_pub_, "Could not start ros control. Power is off!");
+    } else {
+      if (first_ping_error_) {
+        first_ping_error_ = false;
+      } else {
+        ROS_ERROR("Could not ping all devices!");
+        speakError(speak_pub_, "error starting ros control");
+        // check which devices were not pinged successful
+        for (std::pair<std::string, int> &device : dxl_devices) {
+          if (std::find(pinged.begin(), pinged.end(), device.first) != pinged.end()) {
+          } else {
+            ROS_ERROR("%s with id %d missing", device.first.c_str(), device.second);
+          }
+        }
+      }
+    }
+    return false;
+  } else {
+    speakError(speak_pub_, "ross control startup successful");
+    return true;
   }
-  return success;
 }
 
+void threaded_init(std::vector<hardware_interface::RobotHW *> &port_interfaces,
+                   ros::NodeHandle &root_nh,
+                   int &success) {
+  success = true;
+  for (hardware_interface::RobotHW *interface : port_interfaces) {
+    // giving 2 times same node handle to keep interface of base class, dirty
+    success &= interface->init(root_nh, root_nh);
+  }
+}
 
-bool WolfgangHardwareInterface::read()
-{
+bool WolfgangHardwareInterface::init(ros::NodeHandle &root_nh) {
+  // iterate through all ports
+  std::vector<std::thread> threads;
+  std::vector<int *> successes;
+  int i = 0;
+  for (std::vector<hardware_interface::RobotHW *> &port_interfaces : interfaces_) {
+    // iterate through all interfaces on this port
+    int suc = 0;
```

Comment:
Shouldn't this be `bool`?

Reply:
Should normally be a bool, but std::ref() seems to not work with bools, therefore I used an int. Did not find a solution to use a bool. I added a comment to explain, but if you have a better solution, it is very welcome.